### PR TITLE
free: Fix incorrect `CachUse` value with `--line --wide`

### LINKED ISSUE #89: free: incorrect `CachUse` value with `--line --wide`
With procps `free`, the values returned with `--line` and `--line --wide` are very similar:
```
$ free --line
SwapUse           0 CachUse     3406128  MemUse     3489132 MemFree     1997776 
$ free --line --wide
SwapUse           0 CachUse     3406324  MemUse     3481656 MemFree     2005052
```
With uutils `free`, on the other hand, there is a significant decrease in the `CachUse` value with `--line --wide`:
```
$ cargo run free --line
SwapUse           0 CachUse     3415656  MemUse     3492068 MemFree     1992516
$ cargo run free --line --wide
SwapUse           0 CachUse      360556  MemUse     3536136 MemFree     1948428
```

### PR DESCRIPTION
**Pull request by MCredbear**
fix #89

### CODE REVIEW

**Comment by sylvestre** [comment]
please add a test to make sure don't regress, thanks :)

**Comment by MCredbear** [comment]
> please add a test to make sure don't regress, thanks :)

Bro, how do I test this? Run it and GNU's `ps` then paste the result here?

**Comment by Krysztal112233** [comment]
Please note that the community guideline is friendly discussion. This is a small reminder 

Regression testing is necessary to ensure that the issues you fixed don't happen again.

**Comment by sylvestre** [comment]
Please don't call me bro
Tests should be added here:
https://github.com/uutils/procps/blob/main/tests/by-util/test_free.rs

**Comment by MCredbear** [comment]
Sorry, I missed your latest comment, I'll fix it as soon as I can.

**Comment by MCredbear** [comment]
> Please don't call me bro

Also apologize for my poor English, I used to think it is a mannerly call.

**Comment by sylvestre** [comment]
the test is failing:

```

---- test_free::test_line_wide stdout ----
run: /home/runner/work/procps/procps/target/debug/procps free --line
run: /home/runner/work/procps/procps/target/debug/procps free --line --wide
---- test_free::test_line_wide stderr ----
thread 'main' panicked at tests/by-util/test_free.rs:72:5:
assertion failed: `(left == right)`

Diff < left / right > :
<SwapUse           0 CachUse     2572944  MemUse     1143900 MemFree    13026372
>SwapUse           0 CachUse     2572944  MemUse     1144404 MemFree    13025868
```

**Comment by MCredbear** [comment]
> the test is failing:
> 
> ```
> 
> ---- test_free::test_line_wide stdout ----
> run: /home/runner/work/procps/procps/target/debug/procps free --line
> run: /home/runner/work/procps/procps/target/debug/procps free --line --wide
> ---- test_free::test_line_wide stderr ----
> thread 'main' panicked at tests/by-util/test_free.rs:72:5:
> assertion failed: `(left == right)`
> 
> Diff < left / right > :
> <SwapUse           0 CachUse     2572944  MemUse     1143900 MemFree    13026372
> >SwapUse           0 CachUse     2572944  MemUse     1144404 MemFree    13025868
> ```

I think we have to splte the code after `match parse_meminfo()` to another function, otherwise we can't create a reasonable test.

**Comment by Krysztal112233** [comment]
> > the test is failing:
> > 
> > ```
> > 
> > ---- test_free::test_line_wide stdout ----
> > run: /home/runner/work/procps/procps/target/debug/procps free --line
> > run: /home/runner/work/procps/procps/target/debug/procps free --line --wide
> > ---- test_free::test_line_wide stderr ----
> > thread 'main' panicked at tests/by-util/test_free.rs:72:5:
> > assertion failed: `(left == right)`
> > 
> > Diff < left / right > :
> > <SwapUse           0 CachUse     2572944  MemUse     1143900 MemFree    13026372
> > >SwapUse           0 CachUse     2572944  MemUse     1144404 MemFree    13025868
> > ```
> 
> I think we have to splte the code after `match parse_meminfo()` to another function, otherwise we can't create a reasonable test.

No, I think for formatting tests you can write your own algorithm for counting spaces etc.

**Comment by Krysztal112233** [comment]
> the test is failing:
> 
> ```
> 
> ---- test_free::test_line_wide stdout ----
> run: /home/runner/work/procps/procps/target/debug/procps free --line
> run: /home/runner/work/procps/procps/target/debug/procps free --line --wide
> ---- test_free::test_line_wide stderr ----
> thread 'main' panicked at tests/by-util/test_free.rs:72:5:
> assertion failed: `(left == right)`
> 
> Diff < left / right > :
> <SwapUse           0 CachUse     2572944  MemUse     1143900 MemFree    13026372
> >SwapUse           0 CachUse     2572944  MemUse     1144404 MemFree    13025868
> ```

Did this cause the regression test to fail?

**Comment by MCredbear** [comment]
> > the test is failing:
> > ```
> > 
> > ---- test_free::test_line_wide stdout ----
> > run: /home/runner/work/procps/procps/target/debug/procps free --line
> > run: /home/runner/work/procps/procps/target/debug/procps free --line --wide
> > ---- test_free::test_line_wide stderr ----
> > thread 'main' panicked at tests/by-util/test_free.rs:72:5:
> > assertion failed: `(left == right)`
> > 
> > Diff < left / right > :
> > <SwapUse           0 CachUse     2572944  MemUse     1143900 MemFree    13026372
> > >SwapUse           0 CachUse     2572944  MemUse     1144404 MemFree    13025868
> > ```
> 
> Did this cause the regression test to fail?

Theorically, whether it passes is random, every time you run `free` the memory usage might change, so that I think unless we splte the code after `match parse_meminfo()` to another function, otherwise we can't create a reasonable test.

**Comment by Krysztal112233** [comment]
> > > the test is failing:
> > > ```
> > > 
> > > ---- test_free::test_line_wide stdout ----
> > > run: /home/runner/work/procps/procps/target/debug/procps free --line
> > > run: /home/runner/work/procps/procps/target/debug/procps free --line --wide
> > > ---- test_free::test_line_wide stderr ----
> > > thread 'main' panicked at tests/by-util/test_free.rs:72:5:
> > > assertion failed: `(left == right)`
> > > 
> > > Diff < left / right > :
> > > <SwapUse           0 CachUse     2572944  MemUse     1143900 MemFree    13026372
> > > >SwapUse           0 CachUse     2572944  MemUse     1144404 MemFree    13025868
> > > ```
> > 
> > Did this cause the regression test to fail?
> 
> Theorically, whether it passes is random, every time you run `free` the memory usage might change, so that I think unless we splte the code after `match parse_meminfo()` to another function, otherwise we can't create a reasonable test.

I think you are right, just do it :-)

**Comment by MCredbear** [comment]
> > > > the test is failing:
> > > > ```
> > > > 
> > > > ---- test_free::test_line_wide stdout ----
> > > > run: /home/runner/work/procps/procps/target/debug/procps free --line
> > > > run: /home/runner/work/procps/procps/target/debug/procps free --line --wide
> > > > ---- test_free::test_line_wide stderr ----
> > > > thread 'main' panicked at tests/by-util/test_free.rs:72:5:
> > > > assertion failed: `(left == right)`
> > > > 
> > > > Diff < left / right > :
> > > > <SwapUse           0 CachUse     2572944  MemUse     1143900 MemFree    13026372
> > > > >SwapUse           0 CachUse     2572944  MemUse     1144404 MemFree    13025868
> > > > ```
> > > 
> > > 
> > > Did this cause the regression test to fail?
> > 
> > 
> > Theorically, whether it passes is random, every time you run `free` the memory usage might change, so that I think unless we splte the code after `match parse_meminfo()` to another function, otherwise we can't create a reasonable test.
> 
> I think you are right, just do it :-)

Well, I found that it's not enough, the args are paesed before `parse_meminfo()`, but code after `parse_meminfo()` still needs the args, so splting the code after `match parse_meminfo()` to another function will makes the code very ugly.

**Comment by MCredbear** [comment]
Should we break the current struct of code in order to make the proper test?

**Comment by cakebaker** [comment]
I think it is difficult to test it with an integration test. As you said, calling `free` twice might not return the same result.

So I would approach it with unit tests and refactorings.

**Comment by Krysztal112233** [comment]
Any updates? @MCredbear

**Comment by MCredbear** [comment]
No, waiting for @cakebaker 's refactorings...

**Comment by MCredbear** [comment]
Sry, I'll try to refactor this.

**Comment by MCredbear** [comment]
Is there a pretty way to read output from std? Making this test is so awkward...

File: src/uu/free/src/free.rs [hunk 1]
```diff
@@ -280,6 +178,143 @@ pub fn uumain(args: impl uucore::Args) -> UResult<()> {
...
+            }
+
+            if committed {
+                print_committed(mem_info, &n2s);
+            }
+        }
+    }
+}
+
+fn print_one_line(mem_info: &MemInfo, n2s: &dyn Fn(u64) -> String) {
```
Comment (Krysztal112233) [inline src/uu/free/src/free.rs:228]: I think you can made this function just processing the content you wanna print, and print the result outside (such as `uu_main`?)

File: src/uu/free/src/free.rs [hunk 2]
```diff
@@ -280,6 +178,143 @@ pub fn uumain(args: impl uucore::Args) -> UResult<()> {
...
+        "CachUse",
+        n2s(mem_info.buffers + mem_info.cached + mem_info.reclaimable),
+        "MemUse",
+        n2s(mem_info.total - mem_info.available),
+        "MemFree",
+        n2s(mem_info.free)
+    );
+}
+
+fn print_wide(mem_info: &MemInfo, n2s: &dyn Fn(u64) -> String) {
```
Comment (Krysztal112233) [inline src/uu/free/src/free.rs:242]: same

File: src/uu/free/src/free.rs [hunk 3]
```diff
@@ -280,6 +178,143 @@ pub fn uumain(args: impl uucore::Args) -> UResult<()> {
...
+        n2s(mem_info.total - mem_info.available),
+        n2s(mem_info.free),
+        n2s(mem_info.shared),
+        n2s(mem_info.buffers),
+        n2s(mem_info.cached + mem_info.reclaimable),
+        n2s(mem_info.available),
+    );
+}
+
+fn print(mem_info: &MemInfo, n2s: &dyn Fn(u64) -> String) {
```
Comment (Krysztal112233) [inline src/uu/free/src/free.rs:257]: same

File: src/uu/free/src/free.rs [hunk 4]
```diff
@@ -280,6 +178,143 @@ pub fn uumain(args: impl uucore::Args) -> UResult<()> {
...
+        n2s(mem_info.total),
+        n2s(mem_info.total - mem_info.available),
+        n2s(mem_info.free),
+        n2s(mem_info.shared),
+        n2s(mem_info.buffers + mem_info.cached + mem_info.reclaimable),
+        n2s(mem_info.available),
+    );
+}
+
+fn print_lohi(mem_info: &MemInfo, n2s: &dyn Fn(u64) -> String) {
```
Comment (Krysztal112233) [inline src/uu/free/src/free.rs:271]: same

File: src/uu/free/src/free.rs [hunk 5]
```diff
@@ -280,6 +178,143 @@ pub fn uumain(args: impl uucore::Args) -> UResult<()> {
...
+    tuf_combo(
+        "High:",
+        mem_info.high_total,
+        mem_info.high_total - mem_info.high_free,
+        mem_info.high_free.into(),
+        n2s,
+    );
+}
+
+fn print_swap(mem_info: &MemInfo, n2s: &dyn Fn(u64) -> String) {
```
Comment (Krysztal112233) [inline src/uu/free/src/free.rs:288]: same

File: src/uu/free/src/free.rs [hunk 6]
```diff
@@ -280,6 +178,143 @@ pub fn uumain(args: impl uucore::Args) -> UResult<()> {
...
+    tuf_combo(
+        "Swap:",
+        mem_info.swap_total,
+        mem_info.swap_used,
+        mem_info.swap_free.into(),
+        n2s,
+    );
+}
+
+fn print_total(mem_info: &MemInfo, n2s: &dyn Fn(u64) -> String) {
```
Comment (Krysztal112233) [inline src/uu/free/src/free.rs:298]: same

File: src/uu/free/src/free.rs [hunk 7]
```diff
@@ -280,6 +178,143 @@ pub fn uumain(args: impl uucore::Args) -> UResult<()> {
...
+    tuf_combo(
+        "Total:",
+        mem_info.total + mem_info.swap_total,
+        mem_info.total - mem_info.available + mem_info.swap_used,
+        (mem_info.free + mem_info.swap_free).into(),
+        n2s,
+    );
+}
+
+fn print_committed(mem_info: &MemInfo, n2s: &dyn Fn(u64) -> String) {
```
Comment (Krysztal112233) [inline src/uu/free/src/free.rs:308]: same

**Comment by Krysztal112233** [comment]
For the last comment, I suggest you understand the formula `g(f())`, you can set the content processing function to `f(_)` and the printing function to `g(x)`, under this formula the program will be easier to maintain for us :)

**Comment by MCredbear** [comment]
I know, I saw there were functions that directly outputs to std like `tuf_combo`, so that I didn't do `g(f())`. So now I'm going to do this for every output.

File: src/uu/free/src/free.rs [hunk 8]
```diff
@@ -122,7 +123,7 @@ fn parse_meminfo() -> Result<MemInfo, Box<dyn std::error::Error>> {
 
 // print total - used - free combo that is used for everything except memory for now
 // free can be negative if the memory is overcommitted so it has to be signed
-fn tuf_combo<F>(name: &str, total: u64, used: u64, free: i128, f: F)
+fn gen_tuf_combo_str<F>(name: &str, total: u64, used: u64, free: i128, f: F) -> String
```
Comment (Krysztal112233) [inline src/uu/free/src/free.rs:126]: I think `construct` better than `gen`?

```suggestion
fn construct_tuf_combo_str<F>(name: &str, total: u64, used: u64, free: i128, f: F) -> String
```

**Review by Krysztal112233 — APPROVED** [review]

**Comment by cakebaker** [comment]
Thanks!

**Pull request merged**